### PR TITLE
9540: Add warnings when regenerating keys

### PR DESCRIPTION
**Pull request by cooperka**
To-do:
- [x] Rebase after `9546_token_change_email` (view [real diff](https://github.com/thecartercenter/nemo/compare/9546_token_change_email...9540_regenerate_warnings))

### CODE REVIEW

File: config/locales/en/main.yml [hunk 1]
```diff
@@ -2058,6 +2058,9 @@ en:
     change_twilio_auth_token: "Change Auth Token"
     change_api_key: "Change API Key"
     override_code_warning: "Note that the new override code will not work with forms that have already been downloaded. You should record the existing code if this is a problem. Are you sure you want to regenerate the code?"
+    sms_token_warning: "Note that this will invalidate the previous SMS token for this mission. Are you sure you want to regenerate the token?"
+    api_key_warning: "Note that this will invalidate the previous API key for this user. Are you sure you want to regenerate the key?"
+    sms_auth_code_warning: "Note that this will invalidate the previous SMS auth code for this user. Are you sure you want to regenerate the code?"
```
Comment (cooperka) [inline config/locales/en/main.yml]: I went with "for this user" to make it generic for yourself vs. a user you're editing.

Let me know if you want more detail e.g. describe what the actual consequences are like "you need to change the code you use when sending future SMS messages" but that seems hopefully clear from context.

File: app/views/settings/index.html.erb [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
     <%= f.field(:incoming_sms_numbers_str, type: :textarea) %>
 
-    <%= f.regenerable_field(:incoming_sms_token,
+    <%= f.regenerable_field(:incoming_sms_token, confirm: t('setting.sms_token_warning'),
```
Comment (smoyth) [inline app/views/settings/index.html.erb]: Can you plz refactor all in this file to double quotes? And do this as a rule when touching old files?

Comment (smoyth) [reply app/views/settings/index.html.erb]: Can you also please refactor all `using-` as a CSS class (like below line) to not use underscores? So `using-incoming-sms-token`, `using-universal-sms-token`, etc. Trying to slowly get rid of underscores in CSS classes.

File: app/views/users/form.html.erb [hunk 3]
```diff
@@ -48,8 +48,10 @@
       <%= f.field(:password_confirmation, type: :password) %>
     </div>
   <% end %>
-  <%= f.regenerable_field(:api_key) if !@user.new_record? && can?(:regenerate_api_key, @user) %>
-  <%= f.regenerable_field(:sms_auth_code) if !@user.new_record? && can?(:regenerate_sms_auth_code, @user) %>
+  <%= f.regenerable_field(:api_key, confirm: t('setting.api_key_warning')) if
+          !@user.new_record? && can?(:regenerate_api_key, @user) %>
```
Comment (smoyth) [inline app/views/users/form.html.erb]: Thanks for the cleanup. If breaking a one-line `if`, please make it a standard (3-line) `if`. 

Unfortunately we don't have a linter for ERB ...

File: config/locales/en/main.yml [hunk 4]
```diff
@@ -2061,6 +2061,9 @@ en:
     change_twilio_auth_token: "Change Auth Token"
     change_api_key: "Change API Key"
     override_code_warning: "Note that the new override code will not work with forms that have already been downloaded. You should record the existing code if this is a problem. Are you sure you want to regenerate the code?"
+    sms_token_warning: "Note that this will invalidate the previous SMS token for this mission. Are you sure you want to regenerate the token?"
```
Comment (smoyth) [inline config/locales/en/main.yml]: Please insert `All incoming SMSes will be rejected until your SMS gateway settings are updated accordingly.' in the middle.

File: config/locales/en/main.yml (hunk 1 above)
Comment (smoyth) [reply config/locales/en/main.yml]: 👍

**Comment by cooperka** [comment]
Sounds good, I'll plan to pair any time tomorrow for the new tests.

File: spec/features/settings/settings_regenerable_field_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  describe "incoming sms token" do
+    scenario "can regenerate" do
+      visit("/en/m/#{mission.compact_name}/settings")
+      emails = emails_sent_by { expect_token_regenerated(".setting_incoming_sms_token") }
+      expect(emails.length).to eq(1)
+
+      email = emails.first
+      expect(email.body.to_s).to match(/#{mission.name}/)
+      expect(email.body.to_s).not_to match(/translation_missing/)
+    end
```
Comment (smoyth) [inline spec/features/settings/settings_regenerable_field_spec.rb]: Great job!

File: spec/features/settings/settings_regenerable_field_spec.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  before do
+    login(user)
+  end
+
+  describe "override code" do
+    scenario "can regenerate" do
+      visit("/en/m/#{mission.compact_name}/settings")
+      expect_token_generated(".setting_override_code")
+      expect_token_regenerated(".setting_override_code")
```
Comment (smoyth) [inline spec/features/settings/settings_regenerable_field_spec.rb]: Very nice.

File: spec/features/users/user_form_spec.rb [hunk 7]
```diff
@@ -25,11 +25,12 @@
...
       select("Enumerator", from: "user_assignments_attributes_0_role")
       select("Send password reset instructions via email", from: "user_reset_password_method")
-      click_button("Save")
+      emails = emails_sent_by { click_button("Save") }
       expect(page).to have_content("Success: User created successfully")
 
       # Make sure password email url is correct and no missing translations
-      email = ActionMailer::Base.deliveries.last
+      expect(emails.length).to eq(1)
+      email = emails.first
```
Comment (smoyth) [inline spec/features/users/user_form_spec.rb]: Nice refactor.

**Pull request merged**